File: backend/app/models/alert.py
"""The persisted :class:`Alert` model (shape only).

An ``Alert`` is produced from a detector's
:class:`~app.models.candidate_alert.CandidateAlert` and finalised by the Alert
Engine's cooldown/deduplication gate (see ``docs/ALERT_SCHEMA.md`` §2 and
``docs/DETECTION_RULES.md`` §5). Phase 2 defines the schema only; the create/update
lifecycle that populates ``dedup_key``/``occurrence_count``/``last_seen`` and the
AI fields is introduced in Phase 3.
"""

from typing import Annotated

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.enums import AIStatus, Category, Severity, SourceType
from app.models.json_types import (
    Confidence,
    DedupKeyStr,
    FiniteFloat,
    FiniteJsonValue,
    IPStr,
    NonEmptyStr,
    PositiveFiniteFloat,
    UUIDv4Str,
)


class Alert(BaseModel):
    """A persisted alert row (see ``docs/ALERT_SCHEMA.md`` §2)."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    alert_id: UUIDv4Str
    created_at: PositiveFiniteFloat
    detector_id: NonEmptyStr
    detector_version: NonEmptyStr
    category: Category
    severity: Severity
    confidence: Confidence
    src_ip: IPStr | None = None
    dst_ip: IPStr
    window_start: FiniteFloat
    window_end: FiniteFloat
    evidence: dict[str, FiniteJsonValue]
    threshold_snapshot: dict[str, FiniteJsonValue]
    dedup_key: DedupKeyStr
    source_type: SourceType
    occurrence_count: Annotated[int, Field(ge=1)] = 1
    last_seen: PositiveFiniteFloat
    ai_explanation: str | None = None
    ai_status: AIStatus = "none"

    @model_validator(mode="after")
    def _check_time_relationships(self) -> "Alert":
        if self.window_start > self.window_end:
            raise ValueError("window_start must not be after window_end")
        if self.created_at > self.last_seen:
            raise ValueError("created_at must not be after last_seen")
        return self

File: frontend/src/types/alert.ts
/**
 * Domain types mirroring the backend Alert model and enums
 * (see backend/app/models/alert.py, enums.py and docs/ALERT_SCHEMA.md §2).
 *
 * These are compile-time shapes only; untrusted JSON is checked at runtime by
 * the guards in src/validation before it is trusted as one of these types.
 */

export type SourceType = 'synthetic' | 'replay' | 'live'
export type Protocol = 'TCP' | 'UDP' | 'ICMP' | 'OTHER'
export type Severity = 'low' | 'medium' | 'high' | 'critical'
export type Category = 'reconnaissance' | 'dos'
export type AIStatus = 'none' | 'generated' | 'fallback' | 'error'

/** The two V1 detectors; an unknown detector filter is a backend 422. */
export type DetectorId = 'portscan' | 'synflood'

export const SOURCE_TYPES: readonly SourceType[] = ['synthetic', 'replay', 'live']
export const PROTOCOLS: readonly Protocol[] = ['TCP', 'UDP', 'ICMP', 'OTHER']
export const SEVERITIES: readonly Severity[] = ['low', 'medium', 'high', 'critical']
export const CATEGORIES: readonly Category[] = ['reconnaissance', 'dos']
export const AI_STATUSES: readonly AIStatus[] = ['none', 'generated', 'fallback', 'error']
export const DETECTOR_IDS: readonly DetectorId[] = ['portscan', 'synflood']

/** Ordinal ranking used to decide whether a severity escalates (mirrors backend). */
export const SEVERITY_ORDER: Record<Severity, number> = {
  low: 0,
  medium: 1,
  high: 2,
  critical: 3,
}

/** Human-readable labels for the provenance banner (SYNTHETIC / REPLAYED / LIVE-LAB). */
export const SOURCE_TYPE_LABEL: Record<SourceType, string> = {
  synthetic: 'SYNTHETIC',
  replay: 'REPLAYED',
  live: 'LIVE-LAB',
}

/** A persisted alert row (docs/ALERT_SCHEMA.md §2). */
export interface Alert {
  alert_id: string
  created_at: number
  detector_id: string
  detector_version: string
  category: Category
  severity: Severity
  confidence: number
  src_ip: string | null
  dst_ip: string
  window_start: number
  window_end: number
  evidence: Record<string, unknown>
  threshold_snapshot: Record<string, unknown>
  dedup_key: string
  source_type: SourceType
  occurrence_count: number
  last_seen: number
  ai_explanation: string | null
  ai_status: AIStatus
}

/** Response body for GET /api/v1/alerts (docs/API.md §3.1). */
export interface AlertListResponse {
  items: Alert[]
  total: number
  limit: number
  offset: number
}
